### PR TITLE
Add @QueryMap `mapEncoder` attribute

### PR DESCRIPTION
**Pull request by pilak**
* use `mapEncoder` attribute at method level for what encoder to use
* still use builder `QueryMapEncoder` if no attribute specified

----------------------------------------------
Hello,

Working on a project where we are using **FeignBuilder** in a separate library that is used like a client factory. When using some particular DTO request object, with `@QueryMap` annotation, we want to pass through the POJO getters. Unfortunately, we have to change the Builder instance of `QueryMapEncoder`  from `FieldQueryMapEncoder` to `BeanQueryMapEncoder` to do so, and the consequence is that this implementation will be used in every client that will be built by this factory.
Also we are wondering why the choice of this encoder implementation is not closer to the annocation itself.

Here's a PR that makes possible to add the choice of the `QueryMapEncoder` implementation in a `@QueryMap` annotation attribute.

Doing so it is possible to change a QueryMapEncoder only of a specific method in the same Client

### Summary by CodeRabbit

Release Notes:

- New Feature: Introduced a flexible query map encoding strategy in Feign. Users can now customize the encoding of query parameters by choosing between `BEAN`, `FIELD`, and `DEFAULT` encoding strategies.
- Refactor: Updated `BaseBuilder`, `Contract`, `MethodMetadata`, and `RequestTemplateFactoryResolver` classes to support the new query map encoding feature.
- Test: Added new test cases in `FeignTest` to verify the functionality of the new query map encoding feature and its impact on child POJOs altered by getters.
  
Please note that these changes may require updates to your existing code if you are directly using any of the modified classes or methods.

### CODE REVIEW

**Review by velo — APPROVED** [review]

**Comment by velo** [comment]
Can you give project team edit access so I can rebase this branch and get CI rinnung?

**Comment by pilak** [comment]
> Can you give project team edit access so I can rebase this branch and get CI rinnung?

I guess this is done

**Review by velo — APPROVED** [review]

File: core/src/main/java/feign/MethodMetadata.java [hunk 1]
```diff
@@ -109,6 +110,15 @@ public MethodMetadata queryMapIndex(Integer queryMapIndex) {
...
   }
 
+  public QueryMapEncoder queryMapEncoder() {
+    return queryMapEncoder;
+  }
+
+  public MethodMetadata queryMapEncoder(QueryMapEncoder queryMapEncoder) {
+    this.queryMapEncoder = queryMapEncoder;
+    return this;
+  }
```
[the remark this answers is not in the document]
Comment (pilak) [reply core/src/main/java/feign/MethodMetadata.java:120]: queryMapEncoder is nullable

File: core/src/main/java/feign/RequestTemplateFactoryResolver.java [hunk 2]
```diff
@@ -108,26 +108,28 @@ public RequestTemplate create(Object[] argv) {
...
         return (Map<String, Object>) value;
       }
       try {
-        return queryMapEncoder.encode(value);
+        // encode with @QueryMap annotation if exists otherwise with the one from this resolver
+        return queryMapEncoder != null ? queryMapEncoder.encode(value)
+            : this.queryMapEncoder.encode(value);
       } catch (EncodeException e) {
         throw new IllegalStateException(e);
       }
```
[the remark this answers is not in the document]
Comment (pilak) [reply core/src/main/java/feign/RequestTemplateFactoryResolver.java:135]: what can I do here?

**Comment by pilak** [comment]
@velo do I have to resolve rabbitai issue?

**Comment by velo** [comment]
> @velo do I have to resolve rabbitai reviews?

not really.

I'm thinking about removing it, as it's not that useful

**Review by velo — APPROVED** [review]

**Comment by pilak** [comment]
> > @velo do I have to resolve rabbitai reviews?
> 
> not really.
> 
> I'm thinking about removing it, as it's not that useful

yeah awesome but a tiny bit intrusive :smile:

**Pull request merged**